app/code/Magento/Catalog/Model/Product/Option/Value.php

```diff
@@ -190,27 +190,29 @@ public function getProduct()
     public function saveValues()
     {
         foreach ($this->getValues() as $value) {
-            $this->isDeleted(false);
-            $this->setData(
+            $optionValue = clone $this;
```

Comment:
Do you know exactly what values need to be set on this object to save correctly? I am thinking rather than cloning here we add a factory that can create the object as needed. We should also consider the cleanup of these objects once we are done with them.

Reply:
That depends on what type of custom option you added I think. So no, don't think you have a predefined list of values that need to be set unless we have factories for each option type?

Reply:
@JeroenVanLeusden do you think you could update the code to use a factory rather than a clone and see if this covers all the cases?

Reply:
Sure do, hopefully have some time by the end of this week.

Reply:
@dmanners Code underneath makes saving stop altogether. Am I missing something?

See that a `Magento\Catalog\Model\Product\Option\Value\Repository` is not present, might be an idea to add that or include the saving of custom options in `Magento\Catalog\Model\Product\Option\Repository`?

```PHP
foreach ($this->getValues() as $value) {
    $option = $this->getOption();
    $data = array_merge($value, [
        'option_id' => $option->getId(),
        'store_id' => $option->getStoreId()
    ]);
    $optionValue = $this->customOptionValueFactory->create(['data' => $data]);

    if ((bool) $optionValue->getData('is_delete') === true) {
        if ($optionValue->getId()) {
            $optionValue->deleteValues($optionValue->getId());
            $optionValue->delete();
        }
    } else {
        $optionValue->save();
    }
}
```

Reply:
Hi @JeroenVanLeusden I would be tempted to add the repository for just working with the option values. Though https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Model/Product/Option/Repository.php#L146 does deal with saving of option values.

Reply:
@dmanners What are you proposing? Don't know what the preferred way is to get this done.

Reply:
Hi @JeroenVanLeusden sorry for the delay. I would say your suggestion of adding the saving of custom options in `Magento\Catalog\Model\Product\Option\Repository` would be a good option for this.